### PR TITLE
投稿: 面向初学者的 R 语言深度学习教程 庄亮亮

### PR DESCRIPTION
**Pull request by liangliangzhuang**
非常感谢您的PR, 如果您是在为主站投稿, 请将PR的标题改为"投稿:标题+作者的形式",如:
"投稿: 数据通灵术 杜亚磊"
并保留下面的内容.

- 投稿者请注意
  - [x] 主编审核确认接收
  - [x] 主编安排审稿人
  - [x] 审稿、修改

> 至此，投稿部分的工作结束

- 编辑部分工作
  - [ ] 文字编辑做校对工作。需校对文章分类，目前的文章分类包括：**COS访谈，R会议，R语言，推荐文章，新闻动态，机器学习，统计之都，统计图形，统计应用，统计模型，统计计算，统计软件，职业事业**。不得擅自增加分类。
  - [ ] 微信编辑、发布，阅读原文附主站链接，文末附作者介绍
  - [ ] 论坛帖子负责人发帖，粘贴链接，提供 `forum_id` 加入文章
  - [ ] 合并 PR，发布文章

[投稿指南在这里](https://cosx.org/contribute/)，有任何问题，可以直接在PR留言，其他问题请联系: editor@cos.name。

### CODE REVIEW

**Comment by yufree** [comment]
感谢投稿！请问何老师 @hetong007 有时间看下这篇稿件吗？谢谢！

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+首先，对 506 条数据进行划分。随机选择其中的 70% 数据作为训练样本，另外 30% 数据作为测试样本。
+
+```{r}
+# 构建矩阵
+data <- as.matrix(data)
+dimnames(data) <- NULL
+
+# 数据集划分
+set.seed(1234)
+ind <- sample(2, nrow(data), replace = T, prob = c(.7, .3))
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:157]: 这里为什么使用的是`replace = T`呢

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+data %<>% mutate_if(is.factor, as.numeric)
+knitr::kable(head(data[,1:12]))
+```
+
+### 神经网络可视化
+
+在模型搭建前，可以使用 [NeuralNetTools](https://rdrr.io/cran/NeuralNetTools/man/plotnet.html) 包中的 `plotnet()` 函数绘制网络结构。当然也可以使用其他包，例如：[neuralnet](https://github.com/bips-hb/neuralnet)、[tensorboard](https://tensorflow.rstudio.com/tools/tensorboard/tensorboard/)。
+
+```{r results='hide'}
+n <- nnet(medv ~ .,data = data,size = 5)
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:140]: 这里可视化的是`nnet`包的结构，和后文的网络是不相干的吧。后文用keras搭建的网络可以可视化吗？

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+```
+
+总体来看，预测结果还算不错，但是也有一些预测结果和真实值相差甚远。主要原因是，我们没有调整参数来使模型达到最优的效果。读者可以使用 K 折验证的方法来寻找最有的参数，例如：训练轮数，神经网络层数，各层神经元数等。具体案例可以见 《Deep Learning with R》的第 [3.6.4 节](https://github.com/fmmattioni/deep-learning-with-r-notebooks/blob/master/notebooks/3.6-predicting-house-prices.Rmd)。
+
+### 存储/加载模型
+
+为了保存 Keras 模型以供未来使用，使用 `save_model_tf()` 函数保存模型。加载模型使用 `load_model_tf()` 函数。
+
+```{r}
+save_model_tf(object = model, filepath = "model") #保存模型
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:252]: 建议保存的文件都带一个后缀名，用来标识内容属性

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                  optimizer = 'rmsprop',
+                  metrics = 'mae')
+```
+
+### 拟合模型
+
+拟合模型时，RStudio 的 Viewer 会出现：随着迭代变化的损失函数值。如下所示：
+
+```{r}
+mymodel <- model %>%
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:207]: 能给出一个用GPU训练的例子吗？

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+![图1：人工智能、机器学习和深度学习的关系，"来源于：L. Alzubaidi et al. Journal of Big Data (2021)"](images/paste-96170D25.png)
+
+目前，深度学习已经被应用于诸多领域中，例如：医疗保健、社会网络分析、音频和语音处理（识别和增强）、视觉数据处理方法（多媒体数据分析和计算机视觉）、自然语言处理（翻译和句子分类）等。
+
+### 为什么选择 R 语言做深度学习？
+
+随着 R 的不断发展，利用 R 进行深度学习比以往更容易。并且 R 易学易用，不要求很扎实的编程基础，如今它被广泛地应用于机器学习实践和教学中。即使对 R 语言不是很了解的用户也可以通过一些包来搭建深度学习网络。
+
+## 配置工作环境
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:45]: 因为全文介绍的具体代码比较简单，读者读完并不会觉得`nnet`与`keras`或者`tensorflow`相比有什么不同。如果不准备在文章里具体举例，建议作者在开头对比一下这两个新的包比原来R里一些神经网络包的异同点，同时针对这些异同点给出一些进阶学习的资料，方便感兴趣的读者做进一步深入。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+summary(model)
+```
+
+### 编译模型
+
+使用 mse 损失函数 ------均方误差编译网络，即预测和目标之间差异的二次方。这是回归问题广泛使用的损失函数。使用 mae ------平均绝对误差来监控网络。优化器为 rmsprop。
+
+```{r}
+model %>% compile(loss = 'mse',
+                  optimizer = 'rmsprop',
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:198]: `rmsprop`对于这篇文章的目标读者来说应该是一个陌生的概念，建议要么换成比较基础的sgd，要么给一个对`rmsprop`的介绍或者链接。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                  metrics = 'mae')
```
Comment (hetong007) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:199]: 为什么使用`mse`作为loss的时候，要用`mae`作为metric呢

**Review by hetong007 — COMMENTED** [review]
总的来说这篇文章的话题应该对很多人来说都很有意思。我提了一些小的建议，欢迎作者采纳或者讨论。

**Comment by yufree** [comment]
感谢何老师 @hetong007 的审稿！请 @liangliangzhuang 参考意见进行修改或讨论，谢谢！

**Comment by liangliangzhuang** [comment]
感谢何老师 @hetong007 的意见！这两天我好好修改下。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 1 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:157]: 这里是从1，2中有放回抽取一个数，概率分别为（0.7，0.3），一共抽取 `nrow(data)`次。

Comment (hetong007) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:157]: 那训练集和测试集不就有重复的数据吗

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 2 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:140]: 嗯嗯是的，前面使用 `nnet` 包只是想介绍下可视化的方式。用keras搭建的网络可以使用 `TensorBoard` 可视化。这部分我将会更新到文章中。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 3 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:252]: 好的，谢谢！

Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:252]: 测试了一下，该函数保存的是一个文件夹，内部包含变量等信息的子文件，所以不好加后缀名。这里我将 “model” 名换成了 “BostonHousing_model” 以突出是该数据集的结果。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 4 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:207]: 用CPU和GPU训练的代码是一样的。如果前期安装了GPU版本的 TensorFlow， 并且监测到设备中有GPU的话。那么在默认情况下，运算分配给设备时优先使用 GPU 设备。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 6 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:198]: 收到

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 4 above)
Comment (hetong007) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:207]: 可以在前面介绍安装GPU版本的时候加上一段这样的说明

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 7 above)
Comment (hetong007) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:199]: 一般来说，我们都会选用与metric匹配的loss来达到更好的优化效果。作为一个给初学者的教程，我希望在这些细节上少让读者走弯路。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 6 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:198]: 测试了下换成sgd预测效果不是很好，我还是给rmsprop介绍下并加上链接吧。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 5 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd:59]: 好的，已经加入了。

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 7 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 哦哦明白了～

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 1 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 这里的样本值为1和2，只对于这两个数进行抽样重复nrow次。所以得到的（1或2）的向量。training <- data[ind==1,1:13]
test <- data[ind==2, 1:13]有重复训练集和测试集吧。

Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 这里的样本值为1和2，只对于这两个数进行抽样重复nrow次。所以得到的（1或2）的向量。后面才是划分训练集和测试集。
training <- data[ind==1,1:13]
test <- data[ind==2, 1:13]

训练集和测试集没有重复的数据。

Comment (hetong007) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 哦哦，那我明白了

Comment (hetong007) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 哦哦，那我明白了

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd (hunk 7 above)
Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.Rmd]: 何老师，请问是我说的这个意思吗？@hetong007

**Comment by yufree** [comment]
何老师 @hetong007 不知现在的修改是否合适？谢谢！

**Review by hetong007 — APPROVED** [review]
LGTM!

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+categories:
+  - 机器学习
+tags:
+  - 深度学习
+output:
+  blogdown::html_page:
+    toc: true
+link-citations: true
+slug: deep-learning-with-r-for-beginners
+---
```
Comment (yufree) [inline content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html:14]: 423316

**Comment by yufree** [comment]
感谢何老师 @hetong007 的审稿，本篇将进入编辑发布流程，谢谢！

File: content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html (hunk 8 above)
Comment (yufree) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html:14]: 请添加审稿人 何通

Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html:14]: 是我添加到文章吗？还是编辑来？

Comment (yufree) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html:16]: > 是我添加到文章吗？还是编辑来？

编辑来就可以了，谢谢！

Comment (liangliangzhuang) [reply content/post/2022-05-21-deep-learning-with-r-for-beginners/index.html:16]: 我下午擅自主张加了下，不好意思。你们到时候看看合不合适吧。
